In [ ]:
import numpy as np
from sympy import Matrix
import matplotlib.pyplot as plt
import seaborn as sns

1. Creating my Personal Matrix

In [ ]:
# my id 0432310005101001
d1 = 1  # last digit of my student ID
d2 = 0  # second last digit of my student ID

A = np.array([[d1+2, d2+1], [2*d1, d2+2]])
print(A)

[[3 1]
 [2 2]]


● Shape of A

● Determinant

● Rank

● Eigenvalues

● Inverse (only if determinant ≠ 0)

In [7]:
size_A = A.shape

det_A = np.linalg.det(A)

rank_A = np.linalg.matrix_rank(A)

eg_values, eg_vectors = np.linalg.eig(A)

if np.linalg.det(A) != 0:
    inv_A = np.linalg.inv(A)
else:
    inv_A = "Matrix is Singular, cannot compute"

print("\nSize of A: ", size_A)
print("\nDeterminant of A: ", det_A)
print("\nRank of A:", rank_A)
print("\nEigenvalues: ", eg_values)
print("\nInverse of A:\n", inv_A)


Size of A:  (2, 2)

Determinant of A:  4.000000000000001

Rank of A: 2

Eigenvalues:  [4. 1.]

Inverse of A:
 [[ 0.5  -0.25]
 [-0.5   0.75]]


Create a new matrix B by changing exactly one value in A:

● Add +1 to one element of your choice

● You must clearly state which element you changed

In [10]:
B = A
B[0][1] = B[0][1] + 1
print(B)

[[3 2]
 [2 2]]


● Shape of B

● Determinant

● Rank

● Eigenvalues

● Inverse (only if determinant ≠ 0)

In [11]:
size_B = B.shape

det_B = np.linalg.det(B)

rank_B = np.linalg.matrix_rank(B)

eg_values, eg_vectors = np.linalg.eig(B)

if np.linalg.det(B) != 0:
    inv_B = np.linalg.inv(B)
else:
    inv_B = "Matrix is Singular, cannot compute"

print("\nSize of B: ", size_B)
print("\nDeterminant of B: ", det_B)
print("\nRank of B:", rank_B)
print("\nEigenvalues: ", eg_values)
print("\nInverse of B:\n", inv_B)


Size of B:  (2, 2)

Determinant of B:  2.0000000000000004

Rank of B: 2

Eigenvalues:  [4.56155281 0.43844719]

Inverse of B:
 [[ 1.  -1. ]
 [-1.   1.5]]


1. How did the determinant change and why?
    
    → Determinant of A = 4

    → Determinant of B = 2

    → The determinant decreased from 4 to 2

The determinant of a 2×2 matrix:
det = ad−bc

By increasing the value of B[0][1], the subtraction term (bc) increased, which reduced the overall determinant

2. Did the rank change? Explain.

    → Rank of A = 2

    → Rank of B = 2

    the rank did not change

Changing one element was not enough to make rows dependent, so the matrix kept full rank

3. How did the eigenvalues respond to the value change?

    → Eigenvalues:  [4. 1.]

    → Eigenvalues:  [4.56155281 0.43844719]

    Eigenvalues depend on all elements of a matrix

4. Is B easier or harder to invert than A? Why?

    → Determinant of A = 4

    → Determinant of B = 2

A smaller determinant means:

   ▹ The matrix is closer to singular

   ▹ Inverse calculations are more sensitive to numerical error

thats why B is slightly less stable to invert